Blog post: Converting Bicep code to Pulumi (#3408)
pulumi/pulumi-hugo: themes/default/content/blog/converting-bicep-to-pulumi/index.md

|> BicepProgram.parameterizeByResourceGroup
|> Transform.bicepProgramToPulumi
|> Transform.modifyComponentPaths rewriteComponentPath
|> Printer.printProgram
```

You can learn more on the implementation of these functions from the source code on GitHub. I've written it in a way that is easy to follow, understand and contribute to. Most of these functions are fully [unit-tested](https://github.com/Zaid-Ajaj/pulumi-converter-bicep/blob/master/src/Tests/Program.fs) and I am working on adding more tests to cover more of the Bicep language features.

## Bonus converter: ARM to Pulumi

Those who are familiar with the Bicep CLI, know that it has a built-in _decompiler_ that converts ARM templates to Bicep. I thought it would be fun to build another converter that takes ARM templates to Pulumi but without writing any conversion logic for it. Instead, I would decompile the ARM template to Bicep, then use the logic from the Bicep converter to generate the final PCL code. Luckily for me, the decompiler code is embedded in the nuget package [Azure.Bicep.Core](https://www.nuget.org/packages/Azure.Bicep.Core) so I didn't to rely on the existence of the Bicep CLI to decompile the ARM template.

> From all of this, I created another converter: `pulumi-converter-arm` that works exactly like the Bicep converter but takes ARM templates as input. Head over to the [repository](https://github.com/Zaid-Ajaj/pulumi-converter-arm) to learn more about how to install it and how to use it.

Reviewer: Likely need to address how this relates to https://www.pulumi.com/arm2pulumi/.

Author: Added a paragraph ✅ 

> Depending on how well this ARM converter does its job, we might consider deprecating [arm2pulumi](https://www.pulumi.com/arm2pulumi/) in favor of it. However we first need to test it against more templates and iron out the rough edges. It would be the ideal situation because there are virtually zero maintenance costs, any improvements to the Bicep converter would automatically get picked up here. Also improvements to the decompiler from the [Azure.Bicep.Core](https://www.nuget.org/packages/Azure.Bicep.Core) nuget package would benefit the converter as well.

---
title: "Converting Bicep code to Pulumi"

> date: 2023-09-27

Reviewer 2: Need to adjust date for actual publish date. Let's try to get it published Friday.

```suggestion
date: 2023-09-29
```

draft: false
meta_desc: In this article, we will look at a new language converter that takes Bicep code and converts it to any of supported Pulumi languages. 
meta_image: meta.png
authors: ["zaid-ajaj"]
tags: ["community", "bicep", "pulumi", "convert"]
---

> Bicep is DSL developed by Microsoft to simplify the authoring of ARM templates and deploy resources to Azure. Today I will be sharing with you a new Pulumi converter plugin that I have been working on that converts Bicep code to any of the supported Pulumi languages.

Reviewer 2: ```suggestion
Bicep is a DSL developed by Microsoft to simplify the authoring of ARM templates and deploy resources to Azure. Today I will be sharing with you a new Pulumi converter plugin that I have been working on that converts Bicep code to any of the supported Pulumi languages.
```

Author: Thanks a lot for the finding the suggestions 🙏 I've applied those to the blog post accordingly ✅

Once installed, you can use it to convert Bicep code to Pulumi. For example, let's say we have the following Bicep code in a file called `example.bicep` which deploys a storage account:

```bicep
resource storage 'Microsoft.Storage/storageAccounts@2022-09-01' = {
  name: 'storageaccount'
  location: resourceGroup().location
  kind: 'StorageV2'
  sku: {
    name: 'Standard_LRS'
  }
}
```

Navigate to the directory where the file is located and run the following command. For example to generate TypeScript code from the Bicep code above, run:

```bash
pulumi convert --from bicep --language typescript --out output -- --entry example.bicep
```

This command will use the Bicep converter plugin we just installed to convert the code in `example.bicep` to TypeScript and place the generated code in the `output` directory.

You can change `typescript` to any of the supported languages such as `csharp`, `python`, `go`, `java` or `yaml` and you are good to go.

If you don't specify the `--entry` flag, the plugin will look for the first file with the `.bicep` extension in the current directory.

> Let us look the generated code from the example above:

Reviewer 2: ```suggestion
Let us look at the generated code from the example above:
```

    fn::invoke:
      Function: azure-native:resources:getResourceGroup
      Arguments:
        resourceGroupName: ${resourceGroupName}
```

{{% /choosable %}}
{{< /chooser >}}

Notice how the in the original Bicep code, we reference the location of the _implicit_ resource group via `resourceGroup().location`. When the converter generates code for the target language, it replaces this _implicit_ resource group with an explicit one by parameterizing the program by a resource group name and using that resource group when deploying the resources.

## Early days for the converter

Currently the converter supports most of the features of Bicep such as parameters, variables, modules, resources, outputs and a number of built-in functions. That said, the converter plugin is still in its early days and it's not yet tested against many real-world Bicep programs. I am actively working on improving it and making it support most of the Bicep language features. The source code is available on this [GitHub repository](https://github.com/Zaid-Ajaj/pulumi-converter-bicep). If you find any issues or have any feedback, please open an issue on the repository.

## How does it work?

> The converter plugin is a native binary built using dotnet in [F#](https://dotnet.microsoft.com/en-us/languages/fsharp) and new Pulumi Converter SDK. Using the core [Pulumi](https://www.nuget.org/packages/Pulumi) nuget package, we shipped an API that allows you to easily build converter plugins that can extend the functionality of `pulumi convert`.

Reviewer 2: Should we clarify that it's an experimental API?

```suggestion
The converter plugin is a native binary built using dotnet in [F#](https://dotnet.microsoft.com/en-us/languages/fsharp) and new Pulumi Converter SDK. Using the core [Pulumi](https://www.nuget.org/packages/Pulumi) nuget package, we shipped an experimental API that allows you to easily build converter plugins that can extend the functionality of `pulumi convert`.
```

> The converter plugin is a native binary built using dotnet in [F#](https://dotnet.microsoft.com/en-us/languages/fsharp) and new Pulumi Converter SDK. Using the core [Pulumi](https://www.nuget.org/packages/Pulumi) nuget package, we shipped an API that allows you to easily build converter plugins that can extend the functionality of `pulumi convert`.

Reviewer 2: Nit: Let's replace all uses of "dotnet" with ".NET" throughout. It's not usually referred to "dotnet" unless you're specifically referring to the CLI.

- https://github.com/dotnet/runtime#readme
- https://dotnet.microsoft.com/en-us/

```suggestion
The converter plugin is a native binary built using dotnet in [F#](https://dotnet.microsoft.com/en-us/languages/fsharp) and new Pulumi Converter SDK. Using the core [Pulumi](https://www.nuget.org/packages/Pulumi) nuget package, we shipped an API that allows you to easily build converter plugins that can extend the functionality of `pulumi convert`.
```

> Building a language converter plugin involes taking the source language, in this case Bicep, and converting it to the our internal Pulumi Configuration Language or PCL for short. The converter implements a function of the following shape:

Reviewer 2: ```suggestion
Building a language converter plugin involves taking the source language, in this case Bicep, and converting it to the our internal Pulumi Configuration Language or PCL for short. The converter implements a function of the following shape:
```

```fsharp
let convertProgram (request: ConvertProgramRequest): ConvertProgramResponse =
    let sourceDirectory = request.SourceDirectory
    let targetDirectory = request.TargetDirectory
    let args = request.Args
    // conversion logic here
```

> In our case, we read the Bicep code from the in the source directory, convert it to PCL and write out the PCL code in the target directory. Then, Pulumi uses the built-in program generation facilities to take care of the rest of the work of generating the target language code from the PCL code so that we don't have to worry about language-specific details. The converter only knows how to generates PCL.

Reviewer 3: ```suggestion
In our case, we read the Bicep code from the in the source directory, convert it to PCL and write out the PCL code in the target directory. Then, Pulumi uses the built-in program generation facilities to take care of the rest of the work of generating the target language code from the PCL code so that we don't have to worry about language-specific details. The converter only needs to know how to generate PCL.
```

Author: Nice catch ✅

> In our case, we read the Bicep code from the in the source directory, convert it to PCL and write out the PCL code in the target directory. Then, Pulumi uses the built-in program generation facilities to take care of the rest of the work of generating the target language code from the PCL code so that we don't have to worry about language-specific details. The converter only needs to know how to generate PCL.

Reviewer 2: ```suggestion
In our case, we read the Bicep code from the source directory, convert it to PCL and write out the PCL code in the target directory. Then, Pulumi uses the built-in program generation facilities to take care of the rest of the work of generating the target language code from the PCL code so that we don't have to worry about language-specific details. The converter only needs to know how to generate PCL.
```

Now, learning how to setup gRPC servers and implement the `Converter` contract is not an easy task unless you are familiar with it. For dotnet, we shipped the _experimental_ Pulumi converter SDK to the main Pulumi nuget package that makes it extremely easy to build converter plugins.

A bare-bones converter plugin which does nothing looks this in F#:

```fsharp
open Pulumi.Experimental.Converter
open Pulumi.Codegen

let convertProgram (request: ConvertProgramRequest): ConvertProgramResponse =
    // TODO: Implement the conversion logic here
    ConvertProgramResponse.Empty

convertProgram
|> Converter.CreateSimple
|> Converter.Serve
|> Async.AwaitTask
|> Async.RunSynchronously
```

This will spin up a gRPC server that implements the `Converter` contract and will serve it on a random port assigned during startup. The Pulumi converter SDK automatically implements gRPC reflection which allows tools such as [Postman](https://www.postman.com/) to discover the available gRPC services and methods and send example requests to them. See [Postman with gRPC](https://blog.postman.com/postman-now-supports-grpc/).

## Publishing converter plugins

> The easiest way to make your converter plugin available to others is to publish it on GitHub releases. The Pulumi CLI expects a naming convention `pulumi-converter-<language>` and will automatically download the plugin from the latest GitHub release of the repository when given the `--server` flag during `pulumi plugin install converter <language> --server <url>`.

Reviewer 3: Maybe use \<name\> instead of \<language\>? Bit confusing that maybe this is talking about typescript/csharp/etc

Author: Yeah `<name>` is better, adjusted ✅
